secret_store/src/key_server_cluster/admin_sessions/servers_set_change_session.rs

```diff
@@ -1045,148 +1045,205 @@ fn check_nodes_set(all_nodes_set: &BTreeSet<NodeId>, new_nodes_set: &BTreeSet<No
-	use key_server_cluster::{NodeId, SessionId, Error, KeyStorage, DummyKeyStorage};
-	use key_server_cluster::cluster::Cluster;
+	use key_server_cluster::{NodeId, SessionId, Error, KeyStorage, NodeKeyPair, PlainNodeKeyPair};
 	use key_server_cluster::cluster_sessions::ClusterSession;
-	use key_server_cluster::cluster::tests::DummyCluster;
-	use key_server_cluster::generation_session::tests::{MessageLoop as GenerationMessageLoop, Node as GenerationNode, generate_nodes_ids};
+	use key_server_cluster::cluster::tests::MessagesLoop;
+	use key_server_cluster::generation_session::tests::{MessageLoop as GenerationMessageLoop};
+	use key_server_cluster::math;
 	use key_server_cluster::message::Message;
 	use key_server_cluster::admin_sessions::ShareChangeSessionMeta;
-	use key_server_cluster::admin_sessions::share_add_session::tests::check_secret_is_preserved;
 	use key_server_cluster::jobs::servers_set_change_access_job::ordered_nodes_hash;
 	use super::{SessionImpl, SessionParams};
 
-	struct Node {
-		pub cluster: Arc<DummyCluster>,
-		pub key_storage: Arc<DummyKeyStorage>,
-		pub session: SessionImpl,
+	pub trait AdminSessionAdapter<S> {
+		const SIGN_NEW_NODES: bool;
+
+		fn create(
+			meta: ShareChangeSessionMeta,
+			admin_public: Public,
+			all_nodes_set: BTreeSet<NodeId>,
+			ml: &MessagesLoop,
+			idx: usize
+		) -> S;
 	}
 
-	struct MessageLoop {
+	pub struct MessageLoop<S> {
+		pub ml: MessagesLoop,
 		pub admin_key_pair: KeyPair,
 		pub original_key_pair: KeyPair,
+		pub original_key_version: H256,
 		pub all_nodes_set: BTreeSet<NodeId>,
 		pub new_nodes_set: BTreeSet<NodeId>,
 		pub all_set_signature: Signature,
 		pub new_set_signature: Signature,
-		pub nodes: BTreeMap<NodeId, Node>,
+		pub sessions: BTreeMap<NodeId, S>,
 		pub queue: VecDeque<(NodeId, NodeId, Message)>,
 	}
 
-	fn create_session(mut meta: ShareChangeSessionMeta, self_node_id: NodeId, admin_public: Public, all_nodes_set: BTreeSet<NodeId>, cluster: Arc<Cluster>, key_storage: Arc<KeyStorage>) -> SessionImpl {
-		meta.self_node_id = self_node_id;
-		SessionImpl::new(SessionParams {
-			meta: meta,
-			all_nodes_set: all_nodes_set,
-			cluster: cluster,
-			key_storage: key_storage,
-			nonce: 1,
-			admin_public: admin_public,
-			migration_id: None,
-		}).unwrap()
+	impl<S> ::std::fmt::Debug for MessageLoop<S> {
+		fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+			write!(f, "{:?}", self.ml)
+		}
 	}
 
-	fn create_node(meta: ShareChangeSessionMeta, admin_public: Public, all_nodes_set: BTreeSet<NodeId>, node: &GenerationNode) -> Node {
-		for n in &all_nodes_set {
-			node.cluster.add_node(n.clone());
+	struct Adapter;
+
+	impl AdminSessionAdapter<SessionImpl> for Adapter {
+		const SIGN_NEW_NODES: bool = true;
+
+		fn create(
+			mut meta: ShareChangeSessionMeta,
+			admin_public: Public,
+			all_nodes_set: BTreeSet<NodeId>,
+			ml: &MessagesLoop,
+			idx: usize
+		) -> SessionImpl {
+			meta.self_node_id = ml.node_key_pair(idx).public().clone();
```

Comment:
clone on `Copy type`

```suggestion
			meta.self_node_id = *ml.node_key_pair(idx).public();
```